### PR TITLE
[bitnami/mariadb] Feat: Extra configuration using my_extra.cnf (for Helm charts/K8s envs)

### PR DESCRIPTION
**Pull request by fivaldi**
Signed-off-by: Filip Valder <filip.valder@usu.com>

### Description of the change

The change introduces usage of `--defaults-extra-file`. The motivation is: For helm charts, it's not possible to utilize  `extraFlags` to set the options, because this config option has to be _early_ on the command line, just behind `--defaults-file`.

### Benefits

The benefit is convenient override while keeping the default configuration. This is suitable for usage within K8s envs. It's not necessary to maintain the whole configuration, but just the overrides, which simplifies upgrading. In combination with: https://docs.bitnami.com/kubernetes/apps/aspnet-core/administration/deploy-extra-resources/ it's possible to mount the my_extra.cnf file to the `/opt/bitnami/mariadb/conf/my_extra.cnf`.

### Possible drawbacks

There shouldn't be any drawbacks...

### Applicable issues

Didn't found an issue in bitnami/containers or bitnami/charts...

### Additional information

I tested manually / successfully all versions 10.3 -> 10.10 docker builds / runs with the `-v` switch mounting the my_extra.cnf and verified that the extra configuration had effect. The extra configuration I tested was within [mysqld] section to set a utf8mb4 character set, and the mysql database was for all the versions created with this setting (`SHOW CREATE DATABASE mysql`). 

The changes of the corresponding files (from version to version) are identical. You can verify them by e.g. sha256 sum. This information may simplify the pull request review.

Basically just:
```
 bitnami/mariadb/10.10/debian-11/rootfs/opt/bitnami/scripts/libmariadb.sh  | 40 ++++++++++++++++++++++++----------------
 bitnami/mariadb/10.10/debian-11/rootfs/opt/bitnami/scripts/mariadb-env.sh |  1 +
 bitnami/mariadb/10.10/debian-11/rootfs/opt/bitnami/scripts/mariadb/run.sh |  6 ++++--
...
 bitnami/mariadb/README.md                                                 |  6 ++++++
```

### CODE REVIEW

**Comment by fivaldi** [comment]
Here's how this could used with bitnami mariadb Helm chart:

```
primary:
  extraVolumeMounts:
    - name: config-extra
      mountPath: /opt/bitnami/mariadb/conf/my_extra.cnf
      subPath: extra.cnf
      readOnly: true
  extraVolumes:
    - name: config-extra
      configMap:
        name: mariadb-extra
extraDeploy:
  - |
    apiVersion: v1
    kind: ConfigMap
    metadata:
      name: mariadb-extra
      labels: 
        ...
    data:
      extra.cnf: |-
        [mysqld]
        character_set_server=utf8mb4
        collation_server=utf8mb4_unicode_ci

        [client]
        default_character_set=utf8mb4
```

If the MR was accepted, I'd file a subsequent MR for the bitnami/charts/mariadb explaining the usage there.

**Comment by fivaldi** [comment]
Please, is there any status update regarding this pull request? Anything I can help with?

**Comment by CeliaGMqrz** [comment]
Hi @fivaldi, sorry for the delay.

Thanks for your contribution and the detailed information. We are testing these changes. When we have updates, you will receive feedback.

File: bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/libmariadb.sh [hunk 1]
```diff
@@ -265,20 +265,20 @@ mysql_initialize() {
...
+                fi
+            else
+                warn "Could not inject custom configuration for the ${DB_FLAVOR} configuration file '$DB_CONF_DIR/bitnami/my_custom.cnf' because it is not writable."
             fi
-        else
-            warn "Could not inject custom configuration for the ${DB_FLAVOR} configuration file '$DB_CONF_DIR/bitnami/my_custom.cnf' because it is not writable."
         fi
+    else
+        warn "The ${DB_FLAVOR} configuration file '${DB_CONF_FILE}' is not writable. Configurations based on environment variables will not be applied for this file."
     fi
```
Comment (migruiz4) [inline bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/libmariadb.sh:282]: Hi @fivaldi,

Nice catch with the $DB_CONF_FILE write permissions!

Nevertheless, it is not true that write permissions are required for the entire code-block, as there could be the case where a read-only config file already contains the my_custom.cnf include sentence (`!include ${DB_CONF_DIR}/bitnami/my_custom.cnf`) and the entire code block would be ignored.

I would like to suggest a different approach here:

```suggestion
    if is_file_writable "$DB_CONF_FILE"; then
        info "Updating 'my.cnf' with custom configuration"
        mysql_update_custom_config
    else
        warn "The ${DB_FLAVOR} configuration file '${DB_CONF_FILE}' is not writable. Configurations based on environment variables will not be applied for this file."
    fi

    if [[ -f "${DB_CONF_DIR}/my_custom.cnf" ]]; then
        if is_file_writable "${DB_CONF_DIR}/bitnami/my_custom.cnf"; then
            info "Injecting custom configuration 'my_custom.cnf'"
            cat "${DB_CONF_DIR}/my_custom.cnf" > "${DB_CONF_DIR}/bitnami/my_custom.cnf"
            if ! grep --silent "!include ${DB_CONF_DIR}/bitnami/my_custom.cnf" "${DB_CONF_FILE}"; then
                if is_file_writable "$DB_CONF_FILE"; then
                    echo "!include ${DB_CONF_DIR}/bitnami/my_custom.cnf" >> "${DB_CONF_FILE}"
                else
                    warn "Could not inject custom configuration. The ${DB_FLAVOR} configuration file '${DB_CONF_FILE}' is not writable."
                fi
            fi
        else
            warn "Could not inject custom configuration for the ${DB_FLAVOR} configuration file '$DB_CONF_DIR/bitnami/my_custom.cnf' because it is not writable."
        fi
    fi
```

This applies to all branches of mariabd, not only '10.9'

**Review by migruiz4 — CHANGES_REQUESTED** [review]
Hi @fivaldi,

Thank you very much for your contribution! Could you please take a look at my comment regarding the `my_custom.cnf` code block?

File: bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/libmariadb.sh [hunk 2]
```diff
@@ -304,7 +304,8 @@ EOF
         if [[ -z "$DB_REPLICATION_MODE" ]] || [[ "$DB_REPLICATION_MODE" = "master" ]]; then
             if [[ "$DB_REPLICATION_MODE" = "master" ]]; then
                 debug "Starting replication"
-                echo "RESET MASTER;" | debug_execute "$DB_BIN_DIR/mysql" --defaults-file="$DB_CONF_FILE" -N -u root
+                [[ -f "$DB_EXTRA_CONF_FILE" ]] && DEFAULTS_EXTRA_FILE_OPT=--defaults-extra-file="$DB_EXTRA_CONF_FILE"
+                echo "RESET MASTER;" | debug_execute "$DB_BIN_DIR/mysql" --defaults-file="$DB_CONF_FILE" ${DEFAULTS_EXTRA_FILE_OPT:-} -N -u root
```
Comment (migruiz4) [inline bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/libmariadb.sh:308]: To be consistent with the coding style, I would suggest using a `local` env variable instead of `DEFAULTS_EXTRA_FILE_OPT`:

```suggestion
               local -a flags=("--defaults-file=${DB_CONF_FILE}")
               [[ -f "$DB_EXTRA_CONF_FILE" ]] && flags+="--defaults-extra-file=${DB_EXTRA_CONF_FILE}"
                echo "RESET MASTER;" | debug_execute "$DB_BIN_DIR/mysql" "${flags[@]}" -N -u root
```

File: bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/libmariadb.sh [hunk 3]
```diff
@@ -515,6 +518,7 @@ mysql_execute_print_output() {
     local -a args=()
     if [[ -f "$DB_CONF_FILE" ]]; then
         args+=("--defaults-file=${DB_CONF_FILE}")
+        [[ -f "$DB_EXTRA_CONF_FILE" ]] && args+=("--defaults-extra-file=${DB_EXTRA_CONF_FILE}")
     fi
```
Comment (migruiz4) [inline bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/libmariadb.sh:522]: Does `--defaults-extra-file` require `--defaults-file`? If not, I would suggest:
```suggestion
    [[ -f "$DB_CONF_FILE" ]] && args+=("--defaults-file=${DB_CONF_FILE}")
    [[ -f "$DB_EXTRA_CONF_FILE" ]] && args+=("--defaults-extra-file=${DB_EXTRA_CONF_FILE}")
```

File: bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/mariadb/run.sh [hunk 4]
```diff
@@ -17,8 +17,10 @@ set -o pipefail
 # mysqld_safe does not allow logging to stdout/stderr, so we stick with mysqld
 EXEC="${DB_SBIN_DIR}/mysqld"
 
-flags=("--defaults-file=${DB_CONF_DIR}/my.cnf" "--basedir=${DB_BASE_DIR}" "--datadir=${DB_DATA_DIR}" "--socket=${DB_SOCKET_FILE}")
-[[ -z "${DB_PID_FILE:-}" ]] || flags+=("--pid-file=${DB_PID_FILE}")
+flags=("--defaults-file=${DB_CONF_FILE}")
+[[ -f "$DB_EXTRA_CONF_FILE" ]] && flags+=("--defaults-extra-file=${DB_EXTRA_CONF_FILE}")
+flags+=("--basedir=${DB_BASE_DIR}" "--datadir=${DB_DATA_DIR}" "--socket=${DB_SOCKET_FILE}")
+[[ -n "${DB_PID_FILE}" ]] && flags+=("--pid-file=${DB_PID_FILE}")
```
Comment (migruiz4) [inline bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/mariadb/run.sh:23]: Please keep the `"${DB_PID_FILE:-}"`, otherwise the code may fail because of the `set -o nounset`.

```suggestion
[[ -n "${DB_PID_FILE:-}" ]] && flags+=("--pid-file=${DB_PID_FILE}")
```

File: bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/libmariadb.sh (hunk 3 above)
Comment (fivaldi) [reply bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/libmariadb.sh:522]: Anywhere else in the code, there seems to be no decision making as of the `--defaults-file` (it's kinda expected to exist / be set). I wouldn't recommend introducing it in this single place, or - if introduced - should be done on all occurrences (separate MR).

File: bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/libmariadb.sh (hunk 1 above)
Comment (fivaldi) [reply bitnami/mariadb/10.9/debian-11/rootfs/opt/bitnami/scripts/libmariadb.sh:282]: This suggestion is not applicable in GitHub. :(

Can you please send me the patch via some pastebin? 

image

**Comment by fivaldi** [comment]
@migruiz4 Thanks a lot for review and suggestions.

I made some intermediate commits (to be squashed). I left one comment on the `--defaults-file` question/suggestion and I couldn't apply the suggestions for the biggest chunk in GitHub (regarding the write access). Feel free to commit it directly as a separate commit. I'd then patch the other versions.

**Comment by fivaldi** [comment]
@migruiz4 Please, did you have a chance to check my (to-be-squashed) commits and comments?

**Comment by fivaldi** [comment]
@migruiz4 ---v

> @migruiz4 Please, did you have a chance to check my (to-be-squashed) commits and comments?

**Comment by CeliaGMqrz** [comment]
Hi @fivaldi, sorry for the delay, these days there are some colleagues who are on vacation and due to the high demand has delayed the response. We are testing the best way to approach this new functionality. 

Thanks for your comments. You will have an update soon.

**Comment by fivaldi** [comment]
Is there any move here, please?

It's painful that originally a simple code change with great respect to the existing code base (tested on all supported versions) literally spills itself during the review process.... The feature could provide a VERY easy way of overriding configs for docker/k8s envs, by just supplying extra configs (couple of lines), instead of keeping the whole static configuration (tens of lines to be compared/verified during each update...) 

It's very discouraging from any future contribs.. .:(

**Comment by CeliaGMqrz** [comment]
Hello @fivaldi,

Thank you for your patience. We're a small team and our capacity is not too high. This is currently in our radar but the priority is not too high. Note that all changes should be tested on each branch and a breaking change should be avoided. I'm going to continue with the review myself and resolve this as soon as possible. Please could you resolve the conflicts? So we can continue with my colleague's suggestion and I can help you to make the changes.
